pkg/lightning/backend/tidb/tidb.go

```diff
@@ -411,27 +469,45 @@ func (be *tidbBackend) WriteRowsToDB(ctx context.Context, tableName string, colu
 		insertStmt.WriteByte(')')
 	}
 	insertStmt.WriteString(" VALUES")
+	return &insertStmt
+}
 
-	// Note: we are not going to do interpolation (prepared statements) to avoid
-	// complication arise from data length overflow of BIT and BINARY columns
+func (be *tidbBackend) execStmts(ctx context.Context, stmtTasks []stmtTask, batch bool) error {
+	for _, stmtTask := range stmtTasks {
+		for i := 0; i < writeRowsMaxRetryTimes; i++ {
+			stmt := stmtTask.stmt
+			_, err := be.db.ExecContext(ctx, stmt)
 
-	for i, row := range rows {
-		if i != 0 {
-			insertStmt.WriteByte(',')
-		}
-		insertStmt.WriteString(string(row))
-	}
+			failpoint.Inject("mockNonRetryableError", func() {
+				// To mock the non-retryable error for TestWriteRowsErrorDowngrading test
+				err = stderrors.New("mock non-retryable error")
+			})
 
-	// Retry will be done externally, so we're not going to retry here.
-	_, err := be.db.ExecContext(ctx, insertStmt.String())
-	if err != nil && !common.IsContextCanceledError(err) {
-		log.L().Error("execute statement failed", zap.String("stmt", redact.String(insertStmt.String())),
-			zap.Array("rows", rows), zap.Error(err))
+			if err != nil {
+				if !common.IsContextCanceledError(err) {
+					log.L().Error("execute statement failed",
+						zap.Array("rows", stmtTask.rows), zap.String("stmt", redact.String(stmt)), zap.Error(err))
+				}
+				// It's batch mode, just return the error.
+				if batch {
+					return errors.Trace(err)
+				}
+				// Retry the non-batch insert here.
+				if common.IsRetryableError(err) {
+					continue
+				}
+				// TODO: count, record and skip the error.
+				// Just return if any error occurs for now.
+				return errors.Trace(err)
+			}
+			// No error, contine the next stmtTask.
+			break
+		}
```

Comment:
Need to return a error here if max retry times is exceeded.

Reply:
I add a check in the line 504 to make sure the logic won't reach here.

```golang
if common.IsRetryableError(err) && i != writeRowsMaxRetryTimes-1 {
```